### PR TITLE
 #5 feat: Adding support for (de-)serialization schemas 

### LINKED ISSUE #5: Integrate with Flink's schema system for using other formats than String

### PR DESCRIPTION
**Pull request by gunnarmorling**
Fixes #5.

### CODE REVIEW

**Comment by gunnarmorling** [comment]
@rmetzger, that's based on PR #24, you can focus on the second commit for reviewing this PR. Thx!

File: sdk/src/test/java/co/decodable/sdk/pipeline/PurchaseOrder.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright Decodable, Inc.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package co.decodable.sdk.pipeline;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonProperty;
```
Comment (gunnarmorling) [inline sdk/src/test/java/co/decodable/sdk/pipeline/PurchaseOrder.java:10]: I can't describe how much I hate the fact that Flink shades Jackson ;)

File: sdk/src/main/java/co/decodable/sdk/pipeline/DecodableStreamSourceBuilder.java [hunk 2]
```diff
@@ -8,26 +8,31 @@
...
    */
-  DecodableStreamSourceBuilder withStreamId(String streamId);
+  DecodableStreamSourceBuilder<T> withStreamId(String streamId);
 
   /** Specifies the start-up mode to use when reading from the stream. */
-  DecodableStreamSourceBuilder withStartupMode(StartupMode startupMode);
+  DecodableStreamSourceBuilder<T> withStartupMode(StartupMode startupMode);
+
+  /** Specifies the deserialization schema to be used. */
+  DecodableStreamSourceBuilder<T> withDeserializationSchema(
```
Comment (rmetzger) [inline sdk/src/main/java/co/decodable/sdk/pipeline/DecodableStreamSourceBuilder.java:33]: I would distinguish between key and value deserialization schema

File: sdk/src/test/java/co/decodable/sdk/pipeline/PurchaseOrder.java (hunk 1 above)
Comment (rmetzger) [reply sdk/src/test/java/co/decodable/sdk/pipeline/PurchaseOrder.java:10]: You don't have to use Flink's Jackson version. Just add it as a dependency. It is generally a bad idea to rely on Flink shaded stuff, as there's no guarantee that Flink keeps those dependencies across versions.

Comment (rmetzger) [reply sdk/src/test/java/co/decodable/sdk/pipeline/PurchaseOrder.java:10]: Ah ok, you are using Flink's `JsonDeserializationSchema`.

Comment (gunnarmorling) [reply sdk/src/test/java/co/decodable/sdk/pipeline/PurchaseOrder.java:10]: Yes, exactly.

File: sdk/src/main/java/co/decodable/sdk/pipeline/DecodableStreamSourceBuilder.java (hunk 2 above)
Comment (gunnarmorling) [reply sdk/src/main/java/co/decodable/sdk/pipeline/DecodableStreamSourceBuilder.java:33]: Interesting question, do we really want to do this? I don't think we allow any direct access to the key via SQL pipelines (only by designating specific fields of the value as key column(s)), so I'm having some doubts whether we should open this up via custom pipelines. I'd prefer to keep this question for another time.

Comment (rmetzger) [reply sdk/src/main/java/co/decodable/sdk/pipeline/DecodableStreamSourceBuilder.java:33]: Okay, my concern was around forward compatibility -- If we introduce key / value deserialization schemas, then we have to deprecate the `deserializationSchema` methods.

But I think that's fine.

**Review by rmetzger — APPROVED** [review]

File: sdk/src/main/java/co/decodable/sdk/pipeline/DecodableStreamSourceBuilder.java (hunk 2 above)
Comment (gunnarmorling) [reply sdk/src/main/java/co/decodable/sdk/pipeline/DecodableStreamSourceBuilder.java:33]: We're going to do a Beta release, plus all the types are annotated with `@Incubating`, so we'll still be at liberty to change this, even without prior deprecation.

**Pull request merged**